monitoring/mongodb/alerts.yaml

```diff
@@ -103,24 +103,38 @@ groups:
   - alert: ReplicationLagWarning
     expr: |
       label_replace(
-        max by(rs_nm, member_idx) (
-          (
-            max by(pod, rs_nm) (
-              mongodb_rs_members_optimeDate{namespace="${namespace}", pod=~"${service}.*", member_state="PRIMARY"}
+        (
+          max by(rs_nm, member_idx) (
+            (
+              max by(pod, rs_nm) (
+                mongodb_rs_members_optimeDate{namespace="${namespace}", pod=~"${service}.*", member_state="PRIMARY"}
+              )
+              - ignoring(member_idx) group_right min by(pod, rs_nm, member_idx) (
+                mongodb_rs_members_optimeDate{namespace="${namespace}", pod=~"${service}.*", member_state="SECONDARY"}
+              )
+            ) / scalar(
+              max by(rs_nm) (
+                (mongodb_mongod_replset_oplog_head_timestamp{namespace="${namespace}", pod=~"${service}.*"}
+                - on(pod)
+                mongodb_mongod_replset_oplog_tail_timestamp{namespace="${namespace}", pod=~"${service}.*"})
+                * on(pod) group_left()
+                (mongodb_mongod_replset_my_state{namespace="${namespace}", pod=~"${service}.*"} == 1)
+              )
             )
-            - ignoring(member_idx) group_right min by(pod, rs_nm, member_idx) (
-              mongodb_rs_members_optimeDate{namespace="${namespace}", pod=~"${service}.*", member_state="SECONDARY"}
-            )
-          ) / scalar(
-            max by(rs_nm) (
-              (mongodb_mongod_replset_oplog_head_timestamp{namespace="${namespace}", pod=~"${service}.*"}
-              - on(pod)
-              mongodb_mongod_replset_oplog_tail_timestamp{namespace="${namespace}", pod=~"${service}.*"})
-              * on(pod) group_left()
-              (mongodb_mongod_replset_my_state{namespace="${namespace}", pod=~"${service}.*"} == 1)
+          ) * 100 > ${replicationLagWarningThreshold}
+        )
+        AND on(rs_nm)
+        (
+          sum by (rs_nm) (
+            (
+              mongodb_oplog_stats_storageStats_size{namespace="${namespace}", pod=~"${service}.*"}
+              / 
+              mongodb_oplog_stats_storageStats_maxSize{namespace="${namespace}", pod=~"${service}.*"}
             )
-          )
-        ) * 100 > ${replicationLagWarningThreshold},
+            * on(pod) group_left()
+            (mongodb_mongod_replset_my_state{namespace="${namespace}", pod=~"${service}.*"} == 1)
+          ) >= 0.95
+        ),
```

Comment:
We don't necessarily have to have a fully utilized oplog buffer to encounter replication issues right ? In that case, this fixes the issue with false positives at init but might hide lag issues happening early after the creation of the platform. The oplog size is configurable (not sure if we have customers with a larger size than the default) which means that it can be fairly large and can take a while before it's full, which will make the alert trigger much later than it's supposed to.

Reply:
> We don't necessarily have to have a fully utilized oplog buffer to encounter replication issues right ?

No indeed, but product wise, we do not want to alert if there is no risk. The replication lag can be higher during this period with no issue for customers. When it reaches 95% (not the max), then we can start having problems. If a platform was already lagging before the oplog reaches the max storage, then it will need a resync anyway, it's not a replication lag issue anymore.

The oplog size is indeed configurable although we don't change it, so it's 5% (minimum 990MB) of the available storage.

Reply:
A trade-off could be to choose a lower threshold, like 50%, leaving enough room for the startup phases, and some operations. What do you think?

Reply:
I think that works yes, LGTM otherwise